src/Services/OdkLinkService.php

```diff
@@ -726,35 +729,201 @@ private function storeMainSurveyToCustomTable(Xlsform $xlsform, $entry, XlsformT
             $model = new $class;
 
             // check database table existence
-            if (Schema::hasTable($model->getTable())) {
-                // get all column names of a table
-                $columnNames = Schema::getColumnListing($model->getTable());
+            if (!Schema::hasTable($model->getTable())) {
+                return;
             }
 
-            // initialise data array
-            $dataArray = [];
+            // delete previously stored records in this table (if any)
+            $class::where('submission_id', $submissionId)->delete();
+
+            // get data array from main survey
+            $dataArray = $this->prepareDataArray($xlsform, $entry, $section, $schema, $model, $submissionId);
+            // logger($dataArray);
+            // logger('***** ' . $dataArray['farm_id']);
+
+            // create a new database record
+            $class::create($dataArray);
+        }
+    }
+
+
+    // a generic function to extract values from main survey and repeat group entry, returns an array for further processing
+    private function prepareDataArray($xlsform, $entry, $section, $schema, $model, $submissionId): array
+    {
+        // initialise array
+        $result = [];
 
-            // Link the new data model to the current submission
-            $dataArray['submission_id'] = $submissionId;
+        // add submission Id to array
+        $result['submission_id'] = $submissionId;
 
-            // access the value of each ODK variable from a deeply nested array using "dot" notation
-            foreach ($schema as $schemaItem) {
+        // get all column names of a table
+        $columnNames = Schema::getColumnListing($model->getTable());
+
+        // get all foreign key details of a table
+        $foreignKeyDetails = Schema::getForeignKeys($model->getTable());
+        // logger('table: ' . $model->getTable());
+
+        // logger('$foreignKeyDetails: ');
+        // logger($foreignKeyDetails);
+
+        // store foreign key column name and foreign key table name in associative array
+        // TODO: find Laravel array helper function to do the same in a simpler way
+        $foreignKeyColumnNames = [];
+        foreach ($foreignKeyDetails as $foreignKey) {
+            foreach ($foreignKey['columns'] as $foreignKeyColumn) {
+                // logger('foreign_key_column_name: ' . $foreignKeyColumn);
+                $foreignKeyColumnNames[$foreignKeyColumn] = $foreignKey['foreign_table'];
+            }
+            // logger('foreign_table: ' . $foreignKey['foreign_table']);
+        }
```

Comment:
I like the idea of checking for foreign keys on the database. That seems easier than checking for relationships on the data model, which was my first thought.